packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts

```diff
@@ -0,0 +1,58 @@
+import { IntrospectionOutputTypeRef, IntrospectionObjectType } from 'graphql'
+import { flatKind, getObjectType } from './helpers'
+import { LIST, NON_NULL, OBJECT, QueryableField, SCALAR } from './types'
+
+type NestedDict = {
+  [key: string]: NestedDict | undefined
+}
+
+export const nestedFieldsToString = (nestedFields?: NestedDict) => {
+  if (!nestedFields) {
+    return ''
+  }
+
+  const str = Object.entries(nestedFields)
+    .map(([fieldName, field]) => [fieldName, nestedFieldsToString(field)])
+    .flat()
+    .join(' ')
+
+  return `{ ${str} }`
+}
+
+export const queryableFieldToNestedDict = (
+  type: IntrospectionOutputTypeRef,
+  queryableObjectTypes: IntrospectionObjectType[],
+): NestedDict | undefined => {
+  const nestedDict: NestedDict = {}
+
+  if (type.kind === NON_NULL || type.kind === LIST) {
+    return queryableFieldToNestedDict(type.ofType, queryableObjectTypes)
+  }
+
+  if (type.kind === SCALAR) {
+    return undefined
+  }
+
+  if (type.kind === OBJECT) {
+    const objectType = queryableObjectTypes.find(({ name }) => name === type.name)
+    if (objectType) {
+      objectType.fields.forEach((field) => {
+        nestedDict[field.name] = queryableFieldToNestedDict(field.type, queryableObjectTypes)
+      })
+    }
+  }
+
+  return nestedDict
+}
+
+export const getTopLevelFields = (
+  objectTypes: Array<IntrospectionObjectType>,
+  queryName: string,
+): Array<QueryableField> =>
```

Comment:
My old boss used to kill me for not being consistent between `[]` and `Array<>`. I feel like I need to mention it but I really don't want to 😂 I'm not bothered but I can't not mention it now 😱

Reply:
i consistently (i hope!) use `Array<any>` but happy to change